src/stridelayout.jl

```diff
@@ -0,0 +1,300 @@
+    Expr(:call, Expr(:curly, :Contiguous, new_contig))
+end
+
+"""
+contiguous_axis_indicator(::Type{T}) -> Tuple{Vararg{<:Val}}
+
+Returns a tuple boolean `Val`s indicating whether that axis is contiguous.
+"""
+contiguous_axis_indicator(::Type{A}) where {D, A <: AbstractArray{<:Any,D}} = contiguous_axis_indicator(contiguous_axis(A), Val(D))
+contiguous_axis_indicator(::A) where {A <: AbstractArray} = contiguous_axis_indicator(A)
+Base.@pure contiguous_axis_indicator(::Contiguous{N}, ::Val{D}) where {N,D} = ntuple(d -> Val{d == N}(), Val{D}())
+
+"""
+If the contiguous dimension is not the dimension with `Stride_rank{1}`
+"""
+struct ContiguousBatch{N} end
+Base.@pure ContiguousBatch(N::Int) = ContiguousBatch{N}()
+_get(::ContiguousBatch{N}) where {N} = N
+
+"""
+contiguous_batch_size(::Type{T}) -> ContiguousBatch{N}
+
+Returns the size of contiguous batches if `!isone(stride_rank(T, contiguous_axis(T)))`.
+If `isone(stride_rank(T, contiguous_axis(T)))`, then it will return `ContiguousBatch{0}()`.
+If `contiguous_axis(T) == -1`, it will return `ContiguousBatch{-1}()`.
+If unknown, it will return `nothing`.
+"""
+contiguous_batch_size(x) = contiguous_batch_size(typeof(x))
+contiguous_batch_size(::Type) = nothing
+contiguous_batch_size(::Type{Array{T,N}}) where {T,N} = ContiguousBatch{0}()
+contiguous_batch_size(::Type{<:Tuple}) = ContiguousBatch{0}()
+contiguous_batch_size(::Type{<:Union{Transpose{T,A},Adjoint{T,A}}}) where {T,A<:AbstractVecOrMat{T}} = contiguous_batch_size(A)
+contiguous_batch_size(::Type{<:PermutedDimsArray{T,N,I1,I2,A}}) where {T,N,I1,I2,A<:AbstractArray{T,N}} = contiguous_batch_size(A)
+function contiguous_batch_size(::Type{S}) where {N,NP,T,A<:AbstractArray{T,NP},I,S <: SubArray{T,N,A,I}}
+    _contiguous_batch_size(S, contiguous_batch_size(A), contiguous_axis(A))
+end
+_contiguous_batch_size(::Any, ::Any, ::Any) = nothing
+@generated function _contiguous_batch_size(::Type{S}, ::ContiguousBatch{B}, ::Contiguous{C}) where {B,C,N,NP,T,A<:AbstractArray{T,NP},I,S <: SubArray{T,N,A,I}}
+    if I.parameters[C] <: AbstractUnitRange
+        Expr(:call, Expr(:curly, :ContiguousBatch, B))
+    else
+        Expr(:call, Expr(:curly, :ContiguousBatch, -1))
+    end
+end
+
+struct StrideRank{R} end
+Base.@pure StrideRank(R::NTuple{<:Any,Int}) = StrideRank{R}()
+_get(::StrideRank{R}) where {R} = R
+Base.collect(::StrideRank{R}) where {R} = collect(R)
+@inline Base.getindex(::StrideRank{R}, i::Integer) where {R} = R[i]
+@inline Base.getindex(::StrideRank{R}, ::Val{I}) where {R,I} = StrideRank{permute(R, I)}()
+
+function rank_to_sortperm(R::NTuple{N,Int}) where {N}
+    sp = ntuple(zero, Val{N}())
+    r = ntuple(n -> sum(R[n] .≥ R), Val{N}())
+    @inbounds for n in 1:N
+        sp = Base.setindex(sp, n, r[n])
+    end
+    sp
+end
+"""
+rank_to_sortperm(::StrideRank) -> NTuple{N,Int}
+
+Returns the `sortperm` of the stride ranks.
+"""
+@generated Base.sortperm(::StrideRank{R}) where {R} = rank_to_sortperm(R)
```

Comment:
Is this docstring for `rank_to_sortperm` or for `sortperm(::StrideRank)`?

Reply:
I moved it to `rank_to_sortperm`.